security/src/main/java/uk/ac/ox/ndph/mts/security/authorisation/AuthorisationService.java

```diff
@@ -90,56 +69,56 @@ public boolean authorise(String requiredPermission) {
-
-            LOGGER.debug("userId Is - {}", userId);
+            LOGGER.debug("userId Is - {}", getUserId());
             LOGGER.debug("managed identity is - {}", managedIdentity);
-
             //Managed Service Identities represent a call from a service and is
             //therefore authorized.
-            if (securityContextUtil.isInIdentityProviderRole()) {
+            if (securityUtil.isInIdentityProviderRole()) {
                 return true;
             }
 
             // Site IDis should not be null - unless this is init service setting up the root node,
             // but that user is AManagedServiceIdentity
-            if (entitiesSiteIds == null || entitiesSiteIds.stream().anyMatch(Objects::isNull)) {
+            if (siteIds == null || siteIds.stream().anyMatch(Objects::isNull)) {
                 LOGGER.info("SiteID is null therefore request is not unauthorized (permission: {} user: {})",
-                        requiredPermission, userId);
+                        requiredPerm, getUserId());
                 return false;
             }
 
             //get practitioner role assignment
-            List<RoleAssignmentDTO> roleAssignments = practitionerServiceClient.getUserRoleAssignments(userId, token);
+            List<RoleAssignmentDTO> roleAssnmnts =
+                practServClnt.getUserRoleAssignments(getUserId(), getAuthHeaders());
 
-            if (roleAssignments == null || roleAssignments.isEmpty()) {
-                LOGGER.info("User with id {} has no role assignments and therefore is unauthorised.", userId);
+            if (roleAssnmnts == null || roleAssnmnts.isEmpty()) {
+                LOGGER.info("User with id {} has no role assignments and therefore is unauthorised.", getUserId());
                 return false;
             }
 
-            var rolesAssignmentsWithPermission = getRolesAssignmentsWithPermission(requiredPermission, roleAssignments);
-
-            if (rolesAssignmentsWithPermission.isEmpty()) {
+            //Check for the permission in role assignments
+            var roleAssnmntsWtPerm = getRoleAssmntsWtPerm(requiredPerm, roleAssnmnts);
+            if (roleAssnmntsWtPerm.isEmpty()) {
                 return false;
             }
 
-            List<SiteDTO> sites = siteServiceClient.getAssignedSites(
-                                                    SiteServiceClient.bearerAuth(securityContextUtil.getToken()));
+            //If we have required permission at a role and no site given authentication is correct
+            if (siteIds.isEmpty() || siteIds.get(0).isEmpty()) {
+                return true;
+            }
 
-            Set<String> userSites = siteUtil.getUserSites(sites, rolesAssignmentsWithPermission);
+            //Check if the permission exists in roleassignments for siteId
+            if (roleAssnmntsWtPerm.stream().anyMatch(ra -> siteIds.get(0).equalsIgnoreCase(ra.getSiteId()))) {
```

Comment:
This looks to me like it only checks the first site ID in the siteIds parameter? If so need to change the method signature probably (or fix implementation).

Reply:
Hi Alex - after looking into the logic around using a List of siteIds rather than a siteId string - the code is written at the mo to check that if a siteId is null then throw an access is denied error unless its from the initservice which is determied before the error is thrown. Currently we have a create site endpoint that passes the parentSiteId as the siteId, the assign role endpoint that passes the siteId for the user to be assigned to and a host of other endpoints that dont pass any siteId. If any endpoints that dont pass the siteId need to be authzd we create an empty list of siteIds to bypass the null check for the endpoints that should pass a siteId....all this is a little messy to me and also highlights the reason why auth service shouldnt be executing business rules for all services...anyhow if we are to change the current approach  a separate PR will be needed.